### ISSUE TITLE
How to redraw the faux DOM once?

### ISSUE THREAD

**Issue by Ribesg** [Olical/react-faux-dom#120]
There's something I may be missing.

I'm updating my d3 chart exactly like in [this example](https://codesandbox.io/s/JqYGAqlEJ), the only difference being that I'm not using animations.

The reason for not using animations is that they don't work properly on the d3 side, but I'll figure that out at a later point, and it's not related to react-faux-dom.

Meanwhile, I just want to redraw the faux DOM once after updating it. But it looks like you can't do that? I guess it should be a single call to `drawFauxDOM()` but it's not passed to the component props, so you can't call it.

So for now, I'm using `animateFauxDOM(50)` which may redraw up to 3 times. I tried `animateFauxDOM(16)` which didn't work, so I guessed there was something about the CPU speed, and chosed to have a good margin.

Wouldn't a direct way to just update once be better? Did I miss it?

**Comment by Olical** [comment]
Hm, I'm not really sure if you can / how you could do that. I don't fully understand how the animation code works I'm afraid :/ I think a single redraw isn't supported, I think it's built to just operate over a time period. There's another issue (#122) around improving the animation, I'm totally open to changes that make this and deterministic animations possible.

**Comment by Ribesg** [comment]
What's the problem with exposing `drawFauxDOM`?

**Comment by tibotiber** [comment]
@Ribesg `withFauxDOM` was written with animation support as the main target iirc, which could explain why we didn't expose `drawFauxDOM` back then. Seems like exposing it is a good idea and simple to implement. Can you try that on your end, validate that it behaves as you expect and submit a PR? Would be amazing :).

**Comment by Ribesg** [comment]
I'm not familiar enough with the NodeJS environment, I have no idea how to do that.
I mean, I think I know the code changes, but I just don't know how I could test it in my project.

**Comment by Olical** [comment]
I think that #123 will do the trick. Will get it published, let us know :smile:

**Comment by Olical** [comment]
Published in v4.1.0 along with the ability to instantiate new versions of RFD if you need to. May help you if you're finding memory leaks due to reusing one RFD instance for a lot things, thanks to #116.

**Comment by Ribesg** [comment]
I tested v4.1.0.

It works, but I have to use `setTimeout(this.props.drawFauxDOM, 0)` because I'm calling it from my `_updateD3` function which is called in `componentDidUpdate`.

Even using `setTimeout` I'm not sure how to handle the draw as `drawFauxDOM` modifies the state and the `chart` prop, which then calls `componentDidUpdate` again.

It's overall pretty messy and I'm not sure how to handle updates properly.

On this specific issue, maybe the exposed `drawFauxDOM` should actually be a call to `setTimeout`?

**Comment by Olical** [comment]
Hm, that's not ideal. I'm afraid I don't understand what's going on there either. Maybe it'd be best for you to use D3 directly and drop out of the React tree? I end up recommending this to most people that are trying to do anything with animations now. RFD is pretty simple and only really works nicely for very simple use cases.

Personally, I'm okay with that, I'll be the first to admit it's only good for some things and you shouldn't use it for anything moderately complex / animated 😃

**Comment by Ribesg** [comment]
But I'm not doing anything complex. I'm not using animations at all!
The problem is that simple cases like mine *aren't* supported. I just want to draw a chart, then update it.
The only thing simpler than that is drawing a chart and never changing it again, and this isn't supported either: to draw the faux dom to the real dom, you need to start an animation.

**Comment by Olical** [comment]
Huh, okay. My bad, I just wasn't understanding the problem I'm afraid. By the sounds of it, you don't need the mixins at all, right? You can just render your chart and React will re-render when the inputs change? Or maybe I'm misunderstanding again :grimacing:

I'm quite tired at the moment, so I could easily be wrong, it just sounds like you have a simple case that doesn't require the mixins / animation tooling. The normal original usage of RFD will be enough. I hope? I assume there's a reason why you're using the mixin tooling though.

**Comment by Ribesg** [comment]
I'm not sure what are "the mixins", it's my turn not to understand :p

Look at the example given in the readme: https://codesandbox.io/s/yzwyVZjP
This example draws a simple div with some text. Why are we calling something about animation? Where does that 800 value come from? I don't see anything animating?

I don't know what "The normal original usage of RFD" is, I only know that if I update the faux dom without calling `animateFauxDOM`, it never shows up.

**Comment by Olical** [comment]
Oh! So I suspect the root misunderstanding here is that the `withFauxDOM` call is required in the first place. When I first designed RFD it didn't have that, you built your DOM with D3 in your `render` method every time your component was rendered, like this:

> Taken from the readme.

```js
class SomeChart extends React.Component {
  render () {
    // Create your element.
    var el = ReactFauxDOM.createElement('div')

    // Change stuff using actual DOM functions.
    // Even perform CSS selections!
    el.style.setProperty('color', 'red')
    el.setAttribute('class', 'box')

    // Render it to React elements.
    return el.toReact()
  }
}
```

I think that's all you need by the sounds of it, you don't need to call animate etc, that's reserved for those that need to use some of the stateful mutation stuff within D3. If all you're doing is rendering a static chart from some data, the only API you need is:

 * `var el = ReactFauxDOM.createElement('div')`
 * `return el.toReact()`

That was all ever intended to have in the API, the animation part was contributed later to support a whole host of different things that I just didn't bother with because I had no need for them, personally. I _think_ all you need is `toReact()` being returned from your render method.

**Comment by Ribesg** [comment]
Right, this seems closer to what I need, but can you keep the instance (`el` in your example) stored somewhere?
I have static charts that I update periodically, and I almost always update a single `path` element (just its `d` attribute actually) inside a quite complex `svg`, so rebuilding the whole `svg` every time feels wrong.

**Comment by Olical** [comment]
I know what you mean, it seems wasteful, but that is basically how React
works, it's the same deal.

You could build an initial chart on mount, keep it in component local state
and then just mutate a little part of it on render.

I would recommend trying to render it from scratch and just timing it or
checking the performance with devtools. I designed it to render everything
from scratch each time, it's pretty quick :)

If your chart is too big and it's too slow, there are ways to store the
state. I think that's where you need the mixin though, which I've never
personally used in production.

On 13 Nov 2017 21:16, "Gael Ribes" <notifications@github.com> wrote:

Right, this seems closer to what I need, but can you keep the instance (el
in your example) stored somewhere?
I have static charts that I update periodically, and I almost always update
a single path element (just its d attribute actually) inside a quite
complex svg, so rebuilding the whole svg every time feels wrong.

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/Olical/react-faux-dom/issues/120#issuecomment-344060698>,
or mute the thread

**Comment by Olical** [comment]
Just wondering if you solved this? I think the problem was that you were using the animation tooling when you didn't need it?

**Comment by tibotiber** [comment]
@Ribesg sorry I've missed notifications on Github recently 😕. You could also control the re-render in `componentDidUpdate` by only calling `drawFauxDOM` if it needs to be. For example if you'd only need to re-render when props change you'd go like:
```
componentDidUpdate(prevProps, prevState) {
  if (this.props !== prevProps) {
    // call drawFauxDOM here
  }
}
```

**Comment by Ribesg** [comment]
@tibotiber this is basically how I tried to do it at first, but my component has a big props structure and I feel like I would check those props at least two times more than needed. 

In the end I stopped using the whole `withFauxDOM` thing and I'm now recreating the whole faux DOM content every time. 
It works well and I reduced the amount of lines of code I have for this by 40% as I now longer need to 1) have an updatable D3 structure 2) Keep references to D3 shapes, etc. in order to call them again for updates and 3) Have any update code at all.

@Olical My specific problem is solved, but I think there's still something to be done: rework the README so that it's clear that `withFauxDOM` is something secondary to be used with animations only.

I also think that the function exposed in #123 should be clearly documented so that people don't immediately enter an update loop on first usage.
